### ISSUE TITLE
Improve look&feel of the Kiwix Server 404 page

### ISSUE THREAD

**Issue by kelson42** [kiwix/kiwix-tools#710]
The Kiwix Server 404 page is displayed when an HTML page is requested but does not exists in the ZIM file. A 404 error is an HTTP status codethat means the page you were trying to reach on a website couldn't be found on their server.

Today it looks like this:
![Image](https://github.com/user-attachments/assets/a2cbc05e-85b6-44a2-9f59-847d086bc6ab)

This page should obviously localisable.

Here is the link: https://dev.library.kiwix.org/viewer#opentextbooks_en_all_2024-09/openstax.org/subject_does_not_exist

This scenario should happen rarely as ZIM are conceived to remove - as much as possible - the kind of dead links leading to such a page. But this error will always happen time to time.

Today the page is informative but really not fancy. We would like to improve it with a nice design and a Kiwix specific visual:
* Like the the dead Kiwix visual
* We have also a dedicated customised 404 logo

On my opinion we might not put more in emphasis the "404" which is a pure HTTP error and in some cases the readers are not HTTP related, so this might be abusive.

[Here a few very nice 404 pages](https://www.pagecloud.com/blog/best-404-pages)
l

**Comment by siemsie** [comment]
Hi. Can you just clarify your last phrase, you would like a 404 reference, or not? Thanks.

**Comment by kelson42** [comment]
> Hi @kelson42 . Can you just clarify your last phrase, you would like a 404 reference, or not? Thanks.

I suspect that it would be better to not make an explicit reference to the code "404", neither in the text, nor in the visual.

**Comment by siemsie** [comment]
Hi. Here are some first ideas.

![Image](https://github.com/user-attachments/assets/a46d47ba-9890-4196-890a-b07fda81837b)
![Image](https://github.com/user-attachments/assets/3e48bab3-7719-417a-830d-eaad7ca5f68e)

**Comment by kelson42** [comment]
@siemsie Thank you! We will gather feedbacks so we can make an educated chose/response
@juuz0 @veloman-yunkan @rgaudin @Popolechien @sgourdas @benoit74 @ShaopengLin Your feedback is welcome about what is good and what could/should be improved.

**Comment by Popolechien** [comment]
V1 looks nice but why not go with the 404 we already have on our website?
![Image](https://github.com/user-attachments/assets/6412ad86-89ab-4b35-a288-0e220825d58e)

**Comment by benoit74** [comment]
I agree that we must not mention 404, this means nothing to most users.

My preference is for version 2.

I would prefer to revisit a bit wording:
- `The request URL` => `The requested path` or `The requested ZIM item` (with a preference for "path" for simplificity)
- `not found on this server` => `not found in this ZIM`

Regarding the "Make a full text search for ..." paragraph, I feel like it provides way more frustration than value. It is very technical ("full-text search" is a complex term) and very very rarely works from my experience. I would at least replace it with a constant text with something like `You might want to search for this content instead of accessing it via the link you just clicked, maybe this is just a broken link issue and this content is present at another path in the ZIM`. And not automatically suggest search terms which are mostly always bad in my experience.

**Comment by rgaudin** [comment]
- I find the v1 too different from the usual visual of kiwix-serve
- I find both v2 and v3 to not convey the proper meaning.
- I find the 404 to be too tied to understanding what 404 stands for which is probably not the case for a huge share of our user base.

I agree with @benoit74 on the wording

**Comment by siemsie** [comment]
![Image](https://github.com/user-attachments/assets/8421cb28-8654-4144-9718-90307568554d)
These are images we already have (I made v5 with the ?). Could there be something there that works better? 

Either way, need to have consensus on the overall message to find a suitable image. Are we lost? Is it broken?

**Comment by sgourdas** [comment]
While I feel that the 404 is not essential for the every day user to see, I observe that most webpages still choose to display it in some minor way. It seems like everyone understands the general meaning of it, after getting used to seeing it everywhere.

As far for the samples, I am in favor of being consistent with the main website as @Popolechien mentioned, but we should keep the focus on the important details of what happened. In that sense the 404 could be put beside a bigger image of v3, v4/v5 or v6, so they can be used together.

I also think that it is good to provide some getaway resources like links to the home page and more direct and relevant hints like @benoit74 mentioned.

**Comment by kelson42** [comment]
We are working now actively to complete and release `3.8.0` and therefore come back on this. Thank you all for your valuable feedbacks.

@siemsie Can you please propose a white page like v2/v3 but with the v5 logo AND the wording (for both lines) proposed by @benoit74 ?

@rgaudin @benoit74 I think we could implement v3 for 5xx like errors. What do you think?
@rgaudin @benoit74 Should we already request/implement a darkmode version? Or is that too early?

**Comment by rgaudin** [comment]
> [@rgaudin](https://github.com/rgaudin) [@benoit74](https://github.com/benoit74) I think we could implement v3 for 5xx like errors. What do you think? 

Absolutely. Just checked and neither 404 nor 500 responses uses templates. Both deserve a nice handling.

> [@rgaudin](https://github.com/rgaudin) [@benoit74](https://github.com/benoit74) Should we already request/implement a darkmode version? Or is that too early?

Too early IMO given neither the ZIM content nor the homepage support it.

**Comment by benoit74** [comment]
> @rgaudin @benoit74 I think we could implement v3 for 5xx like errors. What do you think?

No opinion

> @rgaudin @benoit74 Should we already request/implement a darkmode version? Or is that too early?

No opinion

**Comment by siemsie** [comment]
Hi all. I hope I understood the request of mixing and matching the previous versions. Here is a v7 and v8. v8 is a rewrite to make shorter and more direct phrases... not sure what I wrote is correct, but can fix that with a v9 if it is the direction we take. 
![Image](https://github.com/user-attachments/assets/6a1b84cb-4239-4ccd-b0aa-8cc8500f4c9b)
![Image](https://github.com/user-attachments/assets/b276363b-f407-438d-a8fa-6b0968be5fa4)

Re dark versions, the Kiwix can't be in white, but a white area can be redefined around the content... let me know if/when it becomes a need.

**Comment by rgaudin** [comment]
Thank you ; I find v8 more readable with the white background and extended text. Feels like effort is given around helping sort out the issue.

Would if be possible to use a single sentence for “The requested path was not found:” and then have it followed by the actual path? I find the cut version weird and it might be difficult to translate.

Also, I think it's important that the path is displayed on a monospace font. Goal of this being to identify typo and such. I don't think with it on its own line the quotes provide any value (can even be misleading).

**Comment by benoit74** [comment]
I second all Renaud points.

**Comment by siemsie** [comment]
Ok here is a v9 taking into account your comments @rgaudin. Thanks for the feedback.
Question: Does it make sense to have a "page not found" following by a "requested path not found"? The top part is a pretty standard phrase which we all recognise (but is effectively a 404) - does it apply?
Thanks.

![Image](https://github.com/user-attachments/assets/1a9db9ec-8211-4a1c-b526-750f17d9ea11)

**Comment by rgaudin** [comment]
LGTM ; I dont mind the repetition. I think the title is more general information than a technical reason (where some subtleties might lie)

**Comment by benoit74** [comment]
LGTM as well, I don't mind the repetition either.

**Comment by siemsie** [comment]
Let me know what you need then and when you need it by, and if there are any further changes. Image, font, sizes and colours? Or just the vector file directly?

**Comment by kelson42** [comment]
As far as I'm concerned, this is ready to implementation. Assigning to @veloman-yunkan 
I should still open the issue for 5xx!

**Comment by veloman-yunkan** [comment]
Is there any source data (HTML, SVG, CSS, etc) that can be used in the implementation?

**Comment by siemsie** [comment]
Hi @veloman-yunkan! Yes, I was just working on it :)

I've packaged the 4xx and 5xx error docs together. I fiddled with some function of an adobe tool to export HTML, I'm not sure it's useful for you, but just in case I popped it in as well. 

Otherwise you will find a PDF with basic instructions for layout and sizes, and the 2 svgs. Don't hesitate to adjust things a little visually if required.

Let me know if you need anything else!

[Kiwix Error 4xx 5xx dev files.zip](https://github.com/user-attachments/files/18655154/Kiwix.Error.4xx.5xx.dev.files.zip)

**Comment by veloman-yunkan** [comment]
Thanks!

**Comment by veloman-yunkan** [comment]
The draft of the new 404 page served for content URLs can be tried in kiwix/libkiwix#1178. I created it based on the PDF document included in the [attachment](https://github.com/user-attachments/files/18655154/Kiwix.Error.4xx.5xx.dev.files.zip) in [@siemsie's comment above](https://github.com/kiwix/kiwix-tools/issues/710#issuecomment-2633386951).

This initial iteration revealed a few problems and gaps in the design:
- The page's responsiveness is poor. The recommended width of the image (25%) doesn't look nice on narrow (phone) screens.
- The PDF contains no information about the width of the white text box. I assumed that it has to match the width of the image.
- The PDF contains no information about vertical spacing between various sections of the page.
- There is a small white region in the top-right corner of the SVG image (above and abutting to the respective corner of Kiwix's monitor). I think it should be transparent.

**Comment by kelson42** [comment]
@siemsie I speak here under your control.

> * The page's responsiveness is poor. The recommended width of the image (25%) doesn't look nice on narrow (phone) screens.

Feel free to adapt the rule because indeed this can not work that way on small screens.

> * The PDF contains no information about the width of the white text box. I assumed that it has to match the width of the image.

Yes, seems a similar challenge.

> * The PDF contains no information about vertical spacing between various sections of the page.

vertically centered, keep a ratio similar to what is in the mockup

> * There is a small white region in the top-right corner of the SVG image (above and abutting to the respective corner of Kiwix's monitor). I think it should be transparent.

Yes, can you fix it or do you need a new version?

**Comment by siemsie** [comment]
Hi, so here are some updated guidelines with more info. I hope I haven't missed anything major?
Indeed there were a couple of issues with the svg, so here it is again.
Drop my any questions you may have, especially for the mobile versions if you're missing anything. If I can receive screenshots it will help me understand what there may be to fix/change.
Thanks.

[Kiwix Server ERROR 4xx 5xx guidelines.pdf](https://github.com/user-attachments/files/18733156/Kiwix.Server.ERROR.4xx.5xx.guidelines.pdf)

![Image](https://github.com/user-attachments/assets/4daaf452-dca7-4d92-8d6b-9898705bf3e3)

**Comment by veloman-yunkan** [comment]
To everyone concerned with fine details in the look-and-feel of error pages. I worked on the HTML/CSS code of the 404 page, trying to make it produce the best match with the provided design, but had to change some numeric values added to the recent guidelines PDF. I feel like it is not the best use of my time since I expect in any case requests for some corrections to whatever version I produce as something that is satisfactory for me. Hence I attach below a self-contained archive of HTML/CSS sources that you can experiment with and edit yourself as you like:

[404_page.zip](https://github.com/user-attachments/files/18800028/404_page.zip)

Once it's ready, I will undertake the further task of integrating it into `kiwix-serve` (the most significant part of which will having the 404 error page internationalized).
